File: src/1425-Constrained-Subsequence-Sum/1425.js
var constrainedSubsetSum = function(nums, k) {
    let q = [], res = -10000, n = nums.length;

    for (let i = 0; i < n; i++) {
        nums[i] = Math.max(nums[i], nums[i] + (q.length ? q[0] : 0));
        while (q.length && nums[i] > q[q.length - 1]) q.pop();
        q.push(nums[i]);

        if (i >= k && q.length && q[0] == nums[i - k]) q.shift();
        res = Math.max(res, nums[i]);
    }
    return res;
};

File: src/1425-Constrained-Subsequence-Sum/1425.go
func constrainedSubsetSum(nums []int, k int) int {
    q := []int{}
    res, n := -10000, len(nums)

    for i := 0; i < n; i++ {
        if len(q) > 0 && q[0] > 0 {
            nums[i] += q[0]
        }
        for len(q) > 0 && nums[i] > q[len(q) - 1] {
            q = q[:len(q) - 1]
        }
        q = append(q, nums[i])
        
        if i >= k && len(q) > 0 && q[0] == nums[i - k] {
            q = q[1:]
        }
        res = max(res, nums[i])
    }
    return res
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}